### PR TITLE
Add Matcher

### PR DESCRIPTION
**Pull request by sachin-101**
## Description

Simple Matcher to match sequences of tokens, based on pattern rules.

```python

def tag_doc_patriotism(matcher, doc, i, matches):
    """Callback function to be passed to matcher."""
    key, start, end = matches[i]
    
    if nlp.vocab.store[key] == "positive":
        doc.set_attribute("patriotism", 1)
    elif nlp.vocab.store[key] == "negative":
        doc.set_attribute("patriotism", -1)

usa_regex = "^[Uu](nited|\.?) ?[Ss](tates|\.?)( ?(of )?[Aa](merica|\.?))?$"
pos_pattern = [
  [{"PRONOUN": True}, {"LEMMA": {"IN" : ["love", "like"]}}, {"LOWER": {"REGEX": usa_regex}}],
  [{"LOWER": "proud"}, {"LOWER": "of"}, {"LOWER": {"REGEX": usa_regex}}]
]
neg_pattern = [{"PRONOUN": True}, {"LOWER": "hate"}, {"LOWER": {"REGEX": usa_regex}}]

matcher = SimpleMatcher(nlp.vocab)

doc1 = nlp("I love USA")
doc2 = nlp("I hate U.S.A.")

matcher.add("positive", [pos_pattern], on_match=tag_doc_patriotism)
matcher.add("negative", [neg_pattern], on_match=tag_doc_patriotism)

matches1 = matcher(doc1)
matches2 = matcher(doc2)

print(doc1.patriotism)  # 1
print(doc2.patriotism)  # -1

# Can iterate through matches and create spans
for match_id, start, end in matches1:
    string_id = nlp.vocab.store[match_id]  # Get string representation
    span = doc[start:end]  # The matched span
    print(match_id, string_id, start, end)
```

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Tests added to `test/local_mode/test_simple_matcher.py`

## Checklist:

* [x] I did follow the [contribution guidelines](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md)
* [x] I have commented my code following [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
* [x] I have added tests for my changes

### CODE REVIEW

**Comment by AlanAboudib** [comment]
@sachin-101 Would you also please create an issue for documenting SimpleMatcher

**Comment by Nilanshrajput** [comment]
@sachin-101  This should wait till ner is done (model and rule based both) i think it might need to be updated after those additions!

File: syfertext/pipeline/matcher.py [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+import syft.serde.msgpack.serde as serde
+
+import re
+from typing import Dict
+from typing import List
+from typing import Union
+
+
+class Matcher:
+    """Match sequences of tokens, based on pattern rules.
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:12]:
```suggestion
    """Match sequences of tokens, based on pattern rules. Inspired by spaCy's `Matcher` class.
```

File: syfertext/pipeline/matcher.py [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+        Args:
+            key (unicode): The match ID.
+
+        Returns:
+            True if the matcher contains the rules for this match ID.
+        """
+
+        return self._normalize_key(key) in self._patterns
+
+    def add(self, key, patterns, on_match=None):
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:47]:
```suggestion
    def add(self, key: str, patterns: list, on_match: callable = None) -> None:
```

File: syfertext/pipeline/matcher.py [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+                 Each dictionary is a single token
+
+        TODO: Avoid the necessity of pattern being a Lists of Lists.
+
+        Args:
+            key (unicode): The match ID.
+            patterns (List): The patterns to add for the given key.
+            on_match (Callable): Optional callback executed on match.
+
+        Returns:
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:80]:
```suggestion
            key: The match ID.
            patterns: The patterns to add for the given key.
            on_match: Optional callback executed on match.
```

File: syfertext/pipeline/matcher.py [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+        """
+
+        # Asserting on match is a callable object
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:83]:
```suggestion
        # Asserting `on_match` is a callable object
```

File: syfertext/pipeline/matcher.py [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+        if on_match is not None and not hasattr(on_match, "__call__"):
+            raise TypeError(f"`on_match` {on_match} is not callable")
+
+        key = self._normalize_key(key)
+        self._patterns[key] = patterns
+        self._preprocessed_patterns[key] = self._process_patterns(patterns)
+        self._callbacks[key] = on_match
+
+    def get(self, key):
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:92]:
```suggestion
    def get(self, key: Union[str, int]) -> Tuple[callable, list]:
```

File: syfertext/pipeline/matcher.py [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+        """ Retrieve the pattern stored for a key.
+
+        Args:
+            key (unicode): The key of the patterns to retrieve.
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:96]:
```suggestion
            key: The key of the patterns to retrieve.
```

File: syfertext/pipeline/matcher.py [hunk 7]
```diff
@@ -0,0 +1,452 @@
+from syfertext.doc import Doc
+from syft.workers.base import BaseWorker
+import syft.serde.msgpack.serde as serde
+
+import re
+from typing import Dict
+from typing import List
+from typing import Union
+
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:9]:
```suggestion
from typing import Tuple
```

File: syfertext/pipeline/matcher.py [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+            key (unicode): The key of the patterns to retrieve.
+
+        Returns:
+            The rule, as an (on_match, patterns) tuple.
+        """
+
+        key = self._normalize_key(key)
+        return self._patterns[key]
+
+    def remove(self, key):
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:105]:
```suggestion
    def remove(self, key: Union[str, int]) -> None:
```

File: syfertext/pipeline/matcher.py [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+        """Remove a rule from the matcher.
+
+        Args:
+            key (unicode): The ID of the match rule.
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:109]:
```suggestion
            key: The ID of the match rule.
```

File: syfertext/pipeline/matcher.py [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+        """
+
+        key = self._normalize_key(key)
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:112]:
```suggestion
        key = self._normalize_key(key)
        
```

File: syfertext/pipeline/matcher.py [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+        try:
+            del self._patterns[key]
+        except KeyError:
+            pass
+
+    def __call__(self, doc):
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:118]:
```suggestion
    def __call__(self, doc: Doc) -> List[Tuple[int, int, int]]:
```

File: syfertext/pipeline/matcher.py [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+        """Find all token sequences matching the supplied pattern.
+
+        Doc must be parsed and it's tokens must have the default attributes.
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:121]:
```suggestion
        Doc must be parsed and its tokens must have the default attributes.
```

File: syfertext/pipeline/matcher.py [hunk 13]
```diff
@@ -0,0 +1,452 @@
...
+            hash value itself.
+        """
+
+        if isinstance(key, str):
+            return self.vocab.store[key]
+        else:
+            return key
+
+    @staticmethod
+    def _process_patterns(patterns):
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:228]:
```suggestion
    def _process_patterns(patterns: List[List[dict]]) -> List[List[tuple]]:
```

File: syfertext/pipeline/matcher.py [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+            matcher.add(key, patterns_, callbacks.get(key, None))
+
+        return matcher
+
+
+# End of Matcher class
+
+
+def _check_match(token_spec, token):
+    """
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:326]: Please add an explanation of what this function is supposed to do. 

(Why don't you put it inside the Matcher class?)

File: syfertext/pipeline/matcher.py [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+        for key, patterns_ in patterns.items():
+            matcher.add(key, patterns_, callbacks.get(key, None))
+
+        return matcher
+
+
+# End of Matcher class
+
+
+def _check_match(token_spec, token):
```
Comment (AlanAboudib) [inline syfertext/pipeline/matcher.py:325]:
```suggestion
def _check_match(token_spec: tuple, token: Token) -> bool:
```

**Comment by AlanAboudib** [comment]
@sachin-101 Lexemes are merged, you can finally continue working on this

**Pull request closed without merging**